modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+// get handled by Audacity
+class InvalidModelCardDocument final : public MessageBoxException
+{
+public:
+   InvalidModelCardDocument(const TranslatableString msg, std::string trace): 
+                           m_msg(msg), 
+                           m_trace(std::move(trace)),
+                            MessageBoxException {
+                               ExceptionType::Internal,
+                               XO("Invalid Model Card Document")
+                            }
+   { 
+   }
+
+   // detailed internal error message
+   virtual const char* what() const throw () 
+   {
+      wxLogError(m_msg.Translation());
+      if (!m_trace.empty()) 
+         wxLogError(wxString(m_trace)); 
+      // TODO: also print the document
+      // TODO: check if document is nullptr
+      return m_msg.Translation().c_str();
+   }
+
+   // user facing message
+   virtual TranslatableString ErrorMessage() const
+      { return XO("Model Card Error: %s")
+               .Format(m_msg);}
+
+   const TranslatableString m_msg;
+   const std::string m_trace;
+};
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &json);
+   Doc ParseFile(const std::string &path);
+}
+
+// validators that accept a default value are no-throw
+namespace validators
+{
+
+   std::string tryGetString(const std::string &key, const Doc& doc);
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue);
+
+   int tryGetInt(const std::string &key, const Doc& doc);
+   int tryGetInt(const std::string &key, const Doc& doc, int defaultValue);
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, const Doc& doc, bool lower=false);
+                                             
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc);
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue);
+
+}
+
+class DeepModelManager;
+
+class ModelCard final
+{
+public: 
+   ModelCard() = default;
+   ModelCard(const ModelCard&) = default;
+
+   // returns {author}/{name}
+   std::string GetRepoID() const;
+
+   bool IsSame(const ModelCard& other) const { return (*this) == other; };
+
+private:
+   friend class DeepModelManager;
+   
+   // throws InvalidModelCardDocument if the given json is not valid. 
+   void DeserializeFromFile(const std::string& path, const Doc& schema);
+   void SerializeToFile(const std::string& path) const;
+
+   template < typename Writer >
+   void Serialize(Writer &writer) const;
+   void Deserialize(const Doc& doc, const Doc& schema);
+
+   void Validate(const Doc& doc, const Doc& schema);
+
+   bool IsLocal() { return m_is_local; }
+   void SetLocal(bool local) { m_is_local = local; }
+
+   std::string GetLocalPath() const { return m_local_path; }
+   void SetLocalPath(const std::string& path) { m_local_path = path; }
+
+public:
+
+   std::string name() const { return m_name; }
+   void name(const std::string &name) { m_name = name; }
```

Comment:
You are inconsistent in the convention for naming of "setter" functions, some with "Set" in the name, some without.

Actually I suggest you use the without convention, and further allow the "chain-call" idiom:

```
ModelCard &name(const std::string &name) { m_name = name; return *this; }
```

And similar, then elsewhere:

```
      card->name(cardName);
      card->author(cardAuthor);
      card->SetLocal(false);
      card->SetLocalPath(audacity::ToUTF8(GetRepoDir(card)));
```

could instead be nicely:

```
(*card).name(cardName)
      .author(cardAuthor)
      .local(false)
      .localPath(audacity::ToUTF8(GetRepoDir(card)));
```

Reply:
resolved in 580b7f943